ios/brave-ios/Sources/BraveVPN/Components/Subscription/BraveVPNInAppPurchaseObserver.swift

```diff
@@ -115,7 +124,63 @@ public class BraveVPNInAppPurchaseObserver: NSObject, SKPaymentTransactionObserv
 
       let errorRestore = SKError(SKError.unknown, userInfo: ["detail": "not-purchased"])
       delegate?.purchaseFailed(error: .transactionError(error: errorRestore))
+      return
+    }
+
+    // Existing Transactions!
+    var callPurchaseDelegateOnce = true
+    let vpnTransactions = queue.transactions.filter({
+      ($0.payment.productIdentifier == BraveVPNProductInfo.ProductIdentifiers.monthlySub)
+        || ($0.payment.productIdentifier == BraveVPNProductInfo.ProductIdentifiers.yearlySub)
+    })
+
+    // There was no VPN purchases
+    if vpnTransactions.isEmpty {
+      let errorRestore = SKError(SKError.unknown, userInfo: ["detail": "not-purchased"])
+      self.delegate?.purchaseFailed(error: .transactionError(error: errorRestore))
+      return
     }
+
+    vpnTransactions
+      .sorted(by: { $0.transactionDate ?? Date() > $1.transactionDate ?? Date() })
+      .forEach { transaction in
+        switch transaction.transactionState {
+        case .purchased:
+          Logger.module.debug("Existing transaction state: purchased")
+          if callPurchaseDelegateOnce {
+            Preferences.VPN.subscriptionProductId.value = transaction.payment.productIdentifier
+            self.delegate?.purchasedOrRestoredProduct(validateReceipt: true)
+          }
+          callPurchaseDelegateOnce = false
+        case .restored:
+          Logger.module.debug("Existing transaction state: restored")
+
+          if callPurchaseDelegateOnce {
+            Preferences.VPN.subscriptionProductId.value = transaction.payment.productIdentifier
+
+            Task {
+              let response = await BraveVPN.validateReceiptData()
+              if response?.status == .expired {
+                self.delegate?.purchaseFailed(error: .receiptError)
+              } else {
+                self.delegate?.purchasedOrRestoredProduct(validateReceipt: false)
+                BraveVPN.clearSkusCredentials(includeExpirationDate: false)
+              }
+            }
+          }
+
+          callPurchaseDelegateOnce = false
+        case .purchasing, .deferred:
+          Logger.module.debug("Received transaction state: purchasing")
+        case .failed:
+          Logger.module.debug("Received transaction state: failed")
+          self.delegate?.purchaseFailed(
+            error: .transactionError(error: transaction.error as? SKError)
```

Comment:
do you not need to check `callPurchaseDelegateOnce` for the failure case?

Reply:
Added.